### PR TITLE
Refactor cookiecutter template identification

### PR DESCRIPTION
**Pull request by michaeljoseph**
In preparation for improving how we handle template references: #385.

### CODE REVIEW

**Comment by pydanny** [comment]
I really like how this breaks the repo code into its own module. Once tests pass I'll start a comprehensive review.

File: cookiecutter/main.py [hunk 1]
```diff
@@ -96,24 +53,13 @@ def cookiecutter(
...
-        # cookiecutters_dir
-        repo_dir = template
-
-    if not os.path.isdir(repo_dir):
-        raise RepositoryNotFound(
-            'The repository {0} could not be located.'.format(template)
-        )
+    repo_dir = determine_repo_dir(
+        template=template,
+        abbreviations=config_dict.get('abbreviations', {}),
```
Comment (hackebrot) [inline cookiecutter/main.py]: ``config_dict`` needs to provide a default value by itself. If it does not currently we need to extend the config module.

Comment (hackebrot) [inline cookiecutter/main.py]: Lets add ``abbreviations`` to ``DEFAULT_CONFIG`` of **config.py**

File: cookiecutter/repository.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+"""Cookiecutter repository functions."""
+from __future__ import unicode_literals
+import os
+import re
+
+from .exceptions import RepositoryNotFound
+from .vcs import clone
+
+BUILTIN_ABBREVIATIONS = {
```
Comment (hackebrot) [inline cookiecutter/repository.py]: As mentioned above it might make sense to have this as a default in **config.py**.

**Comment by hackebrot** [comment]
Great stuff @michaeljoseph! 🍰

**Comment by hackebrot** [comment]
BTW we need to revert the tox changes to test x64 and x86 for appveyor (It does not work anyways with tox as it locates the interpreter by itself and does not use the one with which it has been installed).

Waiting for appveyor to finish is painful.

**Comment by michaeljoseph** [comment]
Thanks @hackebrot, will process the feedback this evening...

**Comment by hackebrot** [comment]
Sure no rush! Have a nice day 😄

File: tests/repository/test_abbreviation_expansion.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        'xx:a', {'xx': '<{0}>'}
+    )
+    assert input_dir == '<a>'
+
+
+def test_abbreviation_expansion_builtin():
+    input_dir = expand_abbreviations(
+        'gh:a', {}
+    )
+    assert input_dir == 'https://github.com/a.git'
```
Comment (luzfcb) [inline tests/repository/test_abbreviation_expansion.py]: This expect a "valid" github url?

Comment (hackebrot) [inline tests/repository/test_abbreviation_expansion.py]: No. Let's change ``gh:a`` to ``gh:pydanny/cookiecutter-django``.

**Comment by michaeljoseph** [comment]
👋 @hackebrot @luzfcb @pydanny

File: cookiecutter/config.py [hunk 4]
```diff
@@ -23,10 +23,16 @@
 
 USER_CONFIG_PATH = os.path.expanduser('~/.cookiecutterrc')
 
+BUILTIN_ABBREVIATIONS = {
+    'gh': 'https://github.com/{0}.git',
+    'bb': 'https://bitbucket.org/{0}',
+}
+
```
Comment (pydanny) [inline cookiecutter/config.py]: Unrelated to this pull request, but I would love to see GitLab added this list. I'll open a new issue.

**Comment by pydanny** [comment]
Very lovely. Quite impressed. Gets my 👍

File: cookiecutter/config.py [hunk 5]
```diff
@@ -23,10 +23,16 @@
...
 DEFAULT_CONFIG = {
     'cookiecutters_dir': os.path.expanduser('~/.cookiecutters/'),
     'replay_dir': os.path.expanduser('~/.cookiecutter_replay/'),
-    'default_context': {}
+    'default_context': {},
+    'abbreviations': BUILTIN_ABBREVIATIONS,
```
Comment (hackebrot) [inline cookiecutter/config.py]: Nice! 👍

**Comment by michaeljoseph** [comment]
❤️

**Pull request merged**